packages/x-charts/src/internals/plugins/featurePlugins/useChartCartesianAxis/createAxisFilterMapper.ts

```diff
@@ -57,43 +63,80 @@ export function createAxisFilterMapper({
-    } else {
-      extremums = getAxisExtremum(axis, direction, seriesConfig, axisIndex, formattedSeries);
+      return createDiscreteScaleGetAxisFilter(axis.data, zoom.start, zoom.end, direction);
     }
 
-    let min: number | Date;
-    let max: number | Date;
+    return createContinuousScaleGetAxisFilter(
+      scaleType,
+      getAxisExtremum(axis, direction, seriesConfig, axisIndex, formattedSeries),
+      zoom.start,
+      zoom.end,
+      direction,
+      axis.data,
+    );
+  };
+}
 
-    const continuousScaleType =
-      !scaleType || scaleType === 'band' || scaleType === 'point' ? 'linear' : scaleType;
+export function createDiscreteScaleGetAxisFilter(
+  axisData: AxisConfig['data'],
+  zoomStart: number,
+  zoomEnd: number,
+  direction: 'x' | 'y',
+): ExtremumFilter {
+  const max = axisData?.length ?? 0;
 
-    [min, max] = getScale(continuousScaleType, extremums, [0, 100]).nice().domain();
+  const minVal = (zoomStart * max) / 100;
+  const maxVal = (zoomEnd * max) / 100;
 
-    min = min instanceof Date ? min.getTime() : min;
-    max = max instanceof Date ? max.getTime() : max;
+  return function filterAxis(value, dataIndex) {
+    const val = value[direction] ?? axisData?.[dataIndex];
 
-    const minVal = min + (zoom.start * (max - min)) / 100;
-    const maxVal = min + (zoom.end * (max - min)) / 100;
+    if (val == null) {
+      // If the value does not exist because of missing data point, or out of range index, we just ignore.
+      return true;
+    }
 
-    return (value, dataIndex) => {
-      const val = value[direction] ?? axis.data?.[dataIndex];
+    return dataIndex > minVal - 1 && dataIndex < maxVal;
+  };
+}
 
-      if (val == null) {
-        // If the value does not exist because of missing data point, or out of range index, we just ignore.
-        return true;
-      }
+export function createContinuousScaleGetAxisFilter(
+  scaleType: ContinuousScaleName | undefined,
+  extrema: readonly [number, number],
+  zoomStart: number,
+  zoomEnd: number,
+  direction: 'x' | 'y',
+  axisData: AxisConfig['data'],
+): ExtremumFilter {
+  let min: number | Date;
+  let max: number | Date;
+
+  [min, max] = getScale(scaleType ?? 'linear', extrema, [0, 100])
+    .nice()
+    .domain();
+
+  min = min instanceof Date ? min.getTime() : min;
+  max = max instanceof Date ? max.getTime() : max;
+
+  const minVal = min + (zoomStart * (max - min)) / 100;
+  const maxVal = min + (zoomEnd * (max - min)) / 100;
+
+  return function filterAxis(value, dataIndex) {
+    const val = value[direction] ?? axisData?.[dataIndex];
+
+    if (val == null) {
+      // If the value does not exist because of missing data point, or out of range index, we just ignore.
+      return true;
+    }
 
-      if (axis.scaleType === 'point' || axis.scaleType === 'band' || typeof val === 'string') {
-        return dataIndex >= minVal && dataIndex <= maxVal;
-      }
+    if (typeof val === 'string') {
```

Comment:
I wonder if we can remove this from continuous scales 🤔 

I kept it because I didn't want to cause a breaking change, but it doesn't make sense for `val` to be a string here if we're using a continuous scale (which only supports numeric values).

If we can remove this, it should simplify this method since we'll no longer need to depend on `axisData`.